data/scenarios/Challenges/Ranching/gated-paddock.yaml

```diff
@@ -0,0 +1,484 @@
+      //    create "boat";
+      //    install self "boat";
+
+          checkIsEnclosedInner;
+          end;
+
+      def boolToInt = \b. if (b) {return 1} {return 0}; end;
+
+      def countAdjacentBlockages =
+
+          turn left;
+          b1 <- isBlockedOrFenced;
+          c1 <- boolToInt b1;
+
+          turn left;
+          b2 <- isBlockedOrFenced;
+          c2 <- boolToInt b2;
+
+          turn left;
+          b3 <- isBlockedOrFenced;
+          c3 <- boolToInt b3;
+
+          turn left;
+          b4 <- isBlockedOrFenced;
+          c4 <- boolToInt b4;
+
+          return $ c1 + c2 + c3 + c4;
+          end;
+
+      // Step forward, observing left and right.
+      def observeLeftAndRight =
+          move;
+          turn left;
+          amBlockedLeft <- isBlockedOrFenced;
+          val1 <- boolToInt amBlockedLeft;
+
+          turn back;
+          amBlockedRight <- isBlockedOrFenced;
+          val2 <- boolToInt amBlockedRight;
+
+          turn right;
+          move;
+          return $ val1 + val2;
+          end;
+
+
+      /** If the four cardinal directions have at most
+      one blockage, then there will exist an orientation
+      where both that direction and its opposite direction
+      are clear.
+      So we can step that direction, check to the left and
+      right of us, then step in the opposite direction
+      and do the same. This allows us to check the 4
+      blocks that touch the corners of the center block.
+      */
+      def countDiagonalBlockages =
+          // First, orient to the clear front-to-back path
+          amBlocked <- isBlockedOrFenced;
+          if amBlocked {turn left;} {};
+
+          // Second, step to both sides
+          fwdCount <- observeLeftAndRight;
+          backCount <- observeLeftAndRight;
+          return $ fwdCount + backCount;
+          end;
+
+      def isStandingOnBridge =
+          onFence <- ishere "fence";
+          onGate <- ishere "gate";
+          if (onFence || onGate) {
+              adjCount <- countAdjacentBlockages;
+              if (adjCount > 1) {
+                  return true;
+              } {
+                  diagCount <- countDiagonalBlockages;
+                  return $ (adjCount + diagCount) > 1;
+              };
+          } {return false};
+          end;
+
+
+      /**
+      There are 3 sheep.
+      They have indices 1, 2, 3.
+      (The base has index 0).
+      */
+      def allSheep = \predicateCmd. \i.
+
+          if (i > 0) {
+              r <- robotnumbered i;
+              didSucceed <- as r {predicateCmd};
+              if didSucceed {
+                  allSheep predicateCmd $ i - 1;
+              } {
+                  return false;
+              };
+          } {
+              return true;
+          }
+          end;
+
+
+      justFilledGap <- as base {
+          isStandingOnBridge;
+      };
+
+      if (justFilledGap) {
+          allSheep checkIsEnclosed 3;
```

Comment:
Well, that is a lot harder now, previously it was enough to save some of the sheep. :confused: 

Now if just one of the sheep drowns, I can no longer win.

Imagine the following scenario:
1. Sheep number 1 randomly runs straight into the ocean
2. The player does not notice
3. Player successfully paddocks all remaining sheep
4. Player waits for the checks to finish
5. The game slows down as it checks the condition every tick
6. Player wonders what is wrong

---

Please do at least one:
1. note what is actually required in the goal description
2. simplify the goal condition to require just `anySheep`
3. move the goal condition to a helper "judge" robot
    - judge robot is loaded once so it can use `run`
    - judge robot could run the condition in `as` on each tick
    - the goal condition would check that judge has "win" or something
    - the judge can interact with the environment, in particular, it can `say` what the player is doing wrong
      - e.g. `say "One sheep has drowned, please start over!"`

Reply:
Also, the judge robot could remember that it already checked the fence connection and not check again on the next tick until the player moves. This would be another speedup. :thinking:

Reply:
Ohh, I ran into exactly this scenario.  One sheep drowned and I did not realize that would prevent me from winning.  What's worse, when I completed the fence but was still sitting on top of the last fence I placed, it pegged my CPU at 100% and the game at 30 ticks per frame, 0.1 fps. =(  It was so sluggish I had to wait like 5-10 seconds just to switch to a different panel in the UI.  As soon as I moved off the fence (with great difficulty, waiting 5-10 seconds in between each UI action) the game suddenly sped up again.

Reply:
I have now relaxed the goal to just save one sheep.

Reply:
> the judge robot could remember

Is there a way to have a persistent "judge" robot that can execute an entire program in one tick?
Also, my judge-algorithm modifies/destroys the environment, so it would have to execute in a "sandboxed" state that is rolled back/discarded upon completion of its program.

What may be useful (and would be facilitated by #795) is if there existed a `god`-capability command that can check if a specific (named) goal has been already achieved.

Reply:
> Is there a way to have a persistent "judge" robot that can execute an entire program in one tick?

We normally use system robots with the `as` command, which is as powerful as you can get.